src/types/redis_stream.cc

```diff
@@ -927,6 +977,145 @@ rocksdb::Status Stream::Range(const Slice &stream_name, const StreamRangeOptions
+  s = storage_->Get(rocksdb::ReadOptions(), stream_cf_handle_, consumer_key, &get_consumer_value);
+  if (!s.ok() && !s.IsNotFound()) {
+    return s;
+  }
+  StreamConsumerMetadata consumer_metadata = decodeStreamConsumerMetadataValue(get_consumer_value);
+  auto now = util::GetTimeStampMS();
+  consumer_metadata.last_idle = now;
+  consumer_metadata.last_active = now;
+
+  if (latest) {
+    options.start = consumergroup_metadata.last_delivered_id;
+    s = range(ns_key, metadata, options, entries);
+    if (!s.ok()) {
+      return s;
+    }
+    StreamEntryID maxid = {0, 0};
+    for (const auto &entry : *entries) {
+      StreamEntryID id;
+      Status st = ParseStreamEntryID(entry.key, &id);
+      if (!st.IsOK()) {
+        return rocksdb::Status::InvalidArgument(st.Msg());
+      }
+      if (id > maxid) {
+        maxid = id;
+      }
+      if (!noack) {
+        std::string pel_key = internalPelKeyFromGroupAndEntryId(ns_key, metadata, group_name, id);
+        StreamPelEntry pel_entry = {0, 0, consumer_name};
+        std::string pel_value = encodeStreamPelEntryValue(pel_entry);
+        batch->Put(stream_cf_handle_, pel_key, pel_value);
+        consumergroup_metadata.entries_read += 1;
+        consumergroup_metadata.pending_number += 1;
+        consumer_metadata.pending_number += 1;
+      }
+    }
+    if (maxid > consumergroup_metadata.last_delivered_id) {
+      consumergroup_metadata.last_delivered_id = maxid;
+    }
+  } else {
+    std::string prefix_key = internalPelKeyFromGroupAndEntryId(ns_key, metadata, group_name, StreamEntryID::Minimum());
+    std::string end_key = internalPelKeyFromGroupAndEntryId(ns_key, metadata, group_name, StreamEntryID::Maximum());
+
+    rocksdb::ReadOptions read_options = storage_->DefaultScanOptions();
+    LatestSnapShot ss(storage_);
+    read_options.snapshot = ss.GetSnapShot();
+    rocksdb::Slice upper_bound(end_key);
+    read_options.iterate_upper_bound = &upper_bound;
+    rocksdb::Slice lower_bound(prefix_key);
+    read_options.iterate_lower_bound = &lower_bound;
+
+    auto iter = util::UniqueIterator(storage_, read_options, stream_cf_handle_);
+    uint64_t count = 0;
+    for (iter->SeekToFirst(); iter->Valid(); iter->Next()) {
+      if (identifySubkeyType(iter->key()) == StreamSubkeyType::StreamPelEntry) {
+        std::string tmp_group_name;
+        StreamEntryID entry_id = groupAndEntryIdFromPelInternalKey(iter->key(), tmp_group_name);
+        if (tmp_group_name != group_name) continue;
+        StreamPelEntry pel_entry = decodeStreamPelEntryValue(iter->value().ToString());
+        if (pel_entry.consumer_name != consumer_name) continue;
+        std::string raw_value;
+        rocksdb::Status st = getEntryRawValue(ns_key, metadata, entry_id, &raw_value);
+        if (!st.ok() && !st.IsNotFound()) {
+          return st;
+        }
+        std::vector<std::string> values;
```

Comment:
What will be if `st.IsNotFound()` is `true`?

Reply:
Is it the case described in the section `What happens when a pending message is deleted?` here https://redis.io/commands/xreadgroup/?

Reply:
Yes, it handles this situation, and i add a new test for it.